src/main/scala/workflow/NodeOptimizationRule.scala

```diff
@@ -5,96 +5,341 @@ import org.apache.spark.rdd.RDD
 /**
  * Node-level optimization, such as selecting a Linear Solver
  *
- * @param sampleFraction The fraction of the RDD to use for operations
+ * @param samplePerPartition The number of items per partition to look at when optimizing nodes
  */
-class NodeOptimizationRule(sampleFraction: Double = 0.01, seed: Long = 0) extends Rule {
+class NodeOptimizationRule(samplePerPartition: Int = 3) extends Rule {
+  case class OptimizationState(
+    optimizedInstructions: Seq[Instruction],
+    unoptimizedInstructionIndexToOptimizedIndex: Int => Int,
+    registers: Map[Int, InstructionOutput],
+    numPerPartitionPerNode: Map[Int, Map[Int, Int]]
+  )
+
+  /**
+   * Get the set of instructions that we need to check for optimizations at.
+   *
+   * This is any instruction that either applies an optimizable transformer,
+   * or fits an optimizable estimator,
+   * and is not supposed to be executed on test data.
+   */
+  private def getInstructionsToOptimize(instructions: Seq[Instruction]): Set[Int] = {
+    instructions.zipWithIndex.collect {
```

Comment:
No, we actually specifically don't want to have a base case so that anything that doesn't match ends up not in the set. Collect is kind of like a mix of map and filter in that sense. If this is too scala magicky let me know and I'll replace it.